modules/lang-painless/src/main/java/org/elasticsearch/painless/toxcontent/UserTreeToXContent.java

```diff
@@ -0,0 +1,689 @@
+        userClassNode.visitChildren(this, scope);
+        builder.endArray();
+
+        end(userClassNode, scope);
+    }
+
+    @Override
+    public void visitFunction(SFunction userFunctionNode, ScriptScope scope) {
+        start(userFunctionNode);
+
+        builder.field("name", userFunctionNode.getFunctionName());
+        builder.field("returns", userFunctionNode.getReturnCanonicalTypeName());
+        if (userFunctionNode.getParameterNames().isEmpty() == false) {
+            builder.field("parameters", userFunctionNode.getParameterNames());
+        }
+        if (userFunctionNode.getCanonicalTypeNameParameters().isEmpty() == false) {
+            builder.field("parameterTypes", userFunctionNode.getCanonicalTypeNameParameters());
+        }
+        builder.field("isInternal", userFunctionNode.isInternal());
+        builder.field("isStatic", userFunctionNode.isStatic());
+        builder.field("isSynthetic", userFunctionNode.isSynthetic());
+        builder.field("isAutoReturnEnabled", userFunctionNode.isAutoReturnEnabled());
+
+        builder.startArray(Fields.BLOCK);
+        userFunctionNode.visitChildren(this, scope);
+        builder.endArray();
+
+        end(userFunctionNode, scope);
+    }
+
+    @Override
+    public void visitBlock(SBlock userBlockNode, ScriptScope scope) {
+        start(userBlockNode);
+
+        builder.startArray("statements");
+        userBlockNode.visitChildren(this, scope);
+        builder.endArray();
+
+        end(userBlockNode, scope);
+    }
+
+    @Override
+    public void visitIf(SIf userIfNode, ScriptScope scope) {
+        start(userIfNode);
+
+        builder.startArray(Fields.CONDITION);
+        userIfNode.getConditionNode().visit(this, scope);
+        builder.endArray();
+
+        block("ifBlock", userIfNode.getIfBlockNode(), scope);
+
+        end(userIfNode, scope);
+    }
+
+    @Override
+    public void visitIfElse(SIfElse userIfElseNode, ScriptScope scope) {
+        start(userIfElseNode);
+
+        builder.startArray(Fields.CONDITION);
+        userIfElseNode.getConditionNode().visit(this, scope);
+        builder.endArray();
+
+        block("ifBlock", userIfElseNode.getIfBlockNode(), scope);
+        block("elseBlock", userIfElseNode.getElseBlockNode(), scope);
+
+        end(userIfElseNode, scope);
+    }
+
+    @Override
+    public void visitWhile(SWhile userWhileNode, ScriptScope scope) {
+        start(userWhileNode);
+        loop(userWhileNode.getConditionNode(), userWhileNode.getBlockNode(), scope);
+        end(userWhileNode, scope);
+    }
+
+    @Override
+    public void visitDo(SDo userDoNode, ScriptScope scope) {
+        start(userDoNode);
+        loop(userDoNode.getConditionNode(), userDoNode.getBlockNode(), scope);
+        end(userDoNode, scope);
+    }
+
+    @Override
+    public void visitFor(SFor userForNode, ScriptScope scope) {
+        start(userForNode);
+
+        // TODO(stu): why is initializerNode ANode instead of an expression
```

Comment:
Don't forget to remove the todo. This has to be an ANode because you can either have an expression or a declaration where declaration is a statement. Kind of frustrating, but both cases must be accounted for.

Reply:
Removed.